lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,297 @@
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+  Returns the embedded data `y`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4,5})
+      iex> seed = 42
+      iex> Scholar.Manifold.TSNE.fit(x, num_components: 2, seed: seed)
+      #Nx.Tensor<
+        f32[4][2]
+        [
+          [1428.49609375, 0.0],
+          [-1499.8349609375, 0.0],
+          [-418.6555480957031, 0.0],
+          [490.010009765625, 0.0]
+        ]
+      >
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(x, seed, opts)
+  end
+
+  defnp fit_n(x, seed, opts \\ []) do
+    {perplexity, learning_rate, num_iters, num_components, exaggeration, init, metric} =
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
+
+          y
+
+        :pca ->
+          Scholar.Decomposition.PCA.fit_transform(x, num_components: num_components)
+      end
+
+    ys = Nx.broadcast(0.0, {num_iters, n, num_components})
+
+    ys = Nx.put_slice(ys, [0, 0, 0], Nx.new_axis(y, 0))
+    ys = Nx.put_slice(ys, [1, 0, 0], Nx.new_axis(y, 0))
+
+    p = p_joint(x, perplexity, metric)
+
+    {y, _, _, _, _} =
+      while {_y = y, ys, learning_rate, p, i = 2}, i < num_iters do
+        q = q_joint(Nx.take(ys, i - 1), metric)
+        grad = gradient(p * exaggeration(i, exaggeration), q, Nx.take(ys, i - 1), metric)
+
+        temp =
+          Nx.take(ys, i - 1) - learning_rate * grad +
+            +momentum(i) * (Nx.take(ys, i - 1) - Nx.take(ys, i - 2))
+
+        ys = Nx.put_slice(ys, [i, 0, 0], Nx.new_axis(temp, 0))
+        {temp, ys, learning_rate, p, i + 1}
+      end
+
+    y
+  end
+
+  defnp pairwise_dist(x, metric) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    t1 = Nx.new_axis(x, 0) |> Nx.broadcast({num_samples, num_samples, num_features})
+    t2 = Nx.new_axis(x, 1) |> Nx.broadcast({num_samples, num_samples, num_features})
+
+    case metric do
+      :squared_euclidean ->
+        Distance.squared_euclidean(t1, t2, axes: [2])
+
+      :euclidean ->
+        Distance.euclidean(t1, t2, axes: [2])
+
+      :manhattan ->
+        Distance.manhattan(t1, t2, axes: [2])
+
+      :cosine ->
+        Distance.cosine(t1, t2, axes: [2])
+
+      :chebyshev ->
+        Distance.chebyshev(t1, t2, axes: [2])
+    end
+  end
+
+  defnp p_conditional(distances, sigmas) do
+    p = Nx.exp(-distances / (2 * Nx.reshape(sigmas, {:auto, 1})) ** 2)
+    {n, _} = Nx.shape(p)
+    p = Nx.put_diagonal(p, Nx.broadcast(0, {n}))
+    p = p + Nx.Constants.epsilon(:f32)
+    p / Nx.sum(p, axes: [1], keep_axes: true)
```

Comment:
Unfortunately, this one doesn't work. You cannot stabilize it so easily like in soft-max by subtracting the max value.

Reply:
Why not?

Reply:
For a simple example, let us assume that `distances` is matrix [[0, 6], [6, 0]], then after subtracting max (6) and setting diagonal to 0, we got zero matrices instead of non-zero.

Reply:
But you can set the diagonal to 0s (or 1s) 8n the end only

Reply:
@msluszniak I ran the following code which seems to emulate your worst-case matrix in 4x4.

```elixir
iex(30)> dist = Nx.subtract(d, Nx.reduce_max(d, axes: [1], keep_axes: true))
#Nx.Tensor<
  s64[4][4]
  [
    [-3, -2, -1, 0],
    [-1, -2, -1, 0],
    [0, -1, -2, -1],
    [0, -1, -2, -3]
  ]
>
iex(31)> p = Nx.exp(Nx.negate(dist) |> Nx.divide(2))
#Nx.Tensor<
  f32[4][4]
  [
    [4.481688976287842, 2.7182817459106445, 1.6487212181091309, 1.0],
    [1.6487212181091309, 2.7182817459106445, 1.6487212181091309, 1.0],
    [1.0, 1.6487212181091309, 2.7182817459106445, 1.6487212181091309],
    [1.0, 1.6487212181091309, 2.7182817459106445, 4.481688976287842]
  ]
>
iex(32)> p_ii = Nx.divide(p, Nx.sum(p, axes: [1], keep_axes: true))
#Nx.Tensor<
  f32[4][4]
  [
    [0.45505422353744507, 0.2760043442249298, 0.16740509867668152, 0.101536326110363],
    [0.23500370979309082, 0.38745561242103577, 0.23500370979309082, 0.1425369530916214],
    [0.1425369530916214, 0.23500370979309082, 0.38745561242103577, 0.23500370979309082],
    [0.101536326110363, 0.16740509867668152, 0.2760043442249298, 0.45505422353744507]
  ]
>
iex(33)> p = Nx.exp(Nx.negate(dist) |> Nx.divide(2)) |> Nx.put_diagonal(Nx.tensor([0, 0, 0, 0]))
#Nx.Tensor<
  f32[4][4]
  [
    [0.0, 2.7182817459106445, 1.6487212181091309, 1.0],
    [1.6487212181091309, 0.0, 1.6487212181091309, 1.0],
    [1.0, 1.6487212181091309, 0.0, 1.6487212181091309],
    [1.0, 1.6487212181091309, 2.7182817459106445, 0.0]
  ]
>
iex(34)> p_ii = Nx.divide(p, Nx.sum(p, axes: [1], keep_axes: true))
#Nx.Tensor<
  f32[4][4]
  [
    [0.0, 0.5064803957939148, 0.30719587206840515, 0.18632373213768005],
    [0.3836517333984375, 0.0, 0.3836517333984375, 0.2326965481042862],
    [0.2326965481042862, 0.3836517333984375, 0.0, 0.3836517333984375],
    [0.18632373213768005, 0.30719587206840515, 0.5064803957939148, 0.0]
  ]
>
iex(35)> dist
#Nx.Tensor<
  s64[4][4]
  [
    [-3, -2, -1, 0],
    [-1, -2, -1, 0],
    [0, -1, -2, -1],
    [0, -1, -2, -3]
  ]
>
iex(36)> p = Nx.exp(Nx.negate(d) |> Nx.divide(2)) |> Nx.put_diagonal(Nx.tensor([0, 0, 0, 0]))
#Nx.Tensor<
  f32[4][4]
  [
    [0.0, 0.6065306663513184, 0.3678794503211975, 0.22313016653060913],
    [0.6065306663513184, 0.0, 0.6065306663513184, 0.3678794503211975],
    [0.3678794503211975, 0.6065306663513184, 0.0, 0.6065306663513184],
    [0.22313016653060913, 0.3678794503211975, 0.6065306663513184, 0.0]
  ]
>
iex(37)> p_ii = Nx.divide(p, Nx.sum(p, axes: [1], keep_axes: true))
#Nx.Tensor<
  f32[4][4]
  [
    [0.0, 0.5064803957939148, 0.30719590187072754, 0.18632373213768005],
    [0.3836517333984375, 0.0, 0.3836517333984375, 0.2326965481042862],
    [0.2326965481042862, 0.3836517333984375, 0.0, 0.3836517333984375],
    [0.18632373213768005, 0.30719590187072754, 0.5064803957939148, 0.0]
  ]
>
```


I still don't get why it doesn't work to stabilize by subtracting the maximum value over each "softmax line"

Reply:
Forgot to show `d`:

```
iex(39)> d
#Nx.Tensor<
  s64[4][4]
  [
    [0, 1, 2, 3],
    [1, 0, 1, 2],
    [2, 1, 0, 1],
    [3, 2, 1, 0]
  ]
>
```

Reply:
Oh, I see, the issue isn't with symmetrical matrices, but with matrices where all non-diagonal elements are equal to the maximum.

We could just return the non-stabilized value iff `all(put_diagonal(d, 0) == 0)`

Reply:
I mean, I run the code on the Iris data set and with the subtracting max it generates incorrect results.

Reply:
my original suggestion was wrong, we have to subtract the max on `distances`, not on `p`

Reply:
Ok, that make sense